File: BOJ/1890. 점프/solution.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const n = +input[0];
const board = input.slice(1).map(v => v.split(' ').map(e => BigInt(e)));

function solution() {
    const dp = Array.from(Array(n), () => Array(n).fill(0n));
    const N = BigInt(n);
    const jump = (x, y) => {
        if (y === N - 1n && x === N - 1n) return 1n;
        if (dp[y][x] !== 0n) return dp[y][x];
        if (board[y][x] === 0n) return 0n;

        const cur = board[y][x];
        if (x + cur < N) dp[y][x] += jump(x + cur, y);
        if (y + cur < N) dp[y][x] += jump(x, y + cur);
        return dp[y][x];
    };

    jump(0n, 0n);
    console.log(dp[0][0].toString());
}

solution();

File: BOJ/1890. 점프/solution.cc
#include <iostream>

using namespace std;

long long board[100][100], dp[100][100];
int n;

long long jump(long long x, long long y) {
    if (y == n - 1 && x == n - 1) return 1;
    if (dp[y][x] != 0) return dp[y][x];
    if (dp[y][x] == 0) return dp[y][x];

    int cur = board[y][x];
    if (x + cur < n) dp[y][x] += jump(x + cur, y);
    if (y + cur < n) dp[y][x] += jump(x, y + cur);
    return dp[y][x];
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    // cout.tie(nullptr); 메모리 초과난다

    cin >> n;
    for (int y = 0; y < n; y += 1) {
        for (int x = 0; x < n; x += 1) {
            cin >> board[y][x];
        }
    }
    
    jump(0, 0);
    cout << dp[0][0];

    return 0;
}